sympy/solvers/ode.py

```diff
@@ -4125,8 +4127,15 @@ def expand_diffx(*args):
         return eq.replace(Derivative, expand_diffx)
 
     # Restore derivatives in solution afterwards
+    from sympy.core.compatibility import PY3
     def unreplace(eq, var):
-        return eq.replace(diffx, lambda e: Derivative(e, var))
+        if PY3:
+            return eq.replace(diffx, lambda e: Derivative(e, var))
+        else:
```

Comment:
I don't understand why this is different between Python 2 and 3.

I added the diffx class and I now realise (seeing the work you've done) that it is fragile in the face of caching. That needs to be fixed and you've added a fix for Python 2 but I'm now worried about how safe it is for Python 3 as well. I'll look into it... (don't let that hold up this PR though)

Reply:
I've seen issues before with classes defined inside of functions. In this case, the class really does have to be defined inside the function, because it is different for each function call. Perhaps it would be better to store `var` and `constant` in the diffx.args. 

If I had to guess, the Python 2-ness of this points to the accursed class registry and the `__cmp__` (see https://github.com/sympy/sympy/pull/13059).

Reply:
Yep:

```
$ cat test.py
from sympy.core.core import all_classes

from sympy import Function, symbols, dsolve, Derivative, Eq

x = symbols('x')
f, g = symbols('f g', cls=Function)

eqn = Eq(Derivative(f(x), x), Derivative(g(x), x))

dsolve(eqn, f(x), hint='nth_algebraic')
dsolve(eqn, f(x), hint='nth_algebraic')

diffxs = [i for i in all_classes if 'diffx' in str(i)]
print(diffxs[0] == diffxs[1])
$ python3 test.py
False
$ python2 test.py
True
```

Because Python 2 uses the ill conceived [`__cmp__` method](https://github.com/sympy/sympy/blob/20234eb7782f1f66c1f1d2a6d05f1179091add0a/sympy/core/core.py#L73) of `BasicMeta` (`__cmp__` does nothing in Python 3).